rest/diagnostic_api.go

```diff
@@ -0,0 +1,341 @@
+				resp.DynamicRoleGrants[roleName][defaultKeyspace] = make(map[string]auth.GrantHistory)
+			}
+			for channel, chanEntry := range role.Channels() {
+				if channel == channels.DocumentStarChannel {
+					continue
+				}
+				grantInfo := auth.GrantHistory{Entries: []auth.GrantHistorySequencePair{{StartSeq: chanEntry.VbSequence.Sequence}}, Source: chanEntry.Source}
+				if roleEntry.VbSequence.Sequence > chanEntry.VbSequence.Sequence {
+					grantInfo.Entries[len(grantInfo.Entries)-1].StartSeq = roleEntry.VbSequence.Sequence
+				}
+				resp.addRoleGrants(roleName, roleEntry.Source, defaultKeyspace, channel, grantInfo)
+			}
+			for channel, chanHistory := range role.ChannelHistory() {
+				if roleEntry.VbSequence.Sequence > chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq {
+					chanHistory.Entries[len(chanHistory.Entries)-1].StartSeq = roleEntry.VbSequence.Sequence
+				}
+				resp.addRoleGrants(roleName, roleEntry.Source, defaultKeyspace, channel, chanHistory)
+			}
+		}
+		for roleName, roleHist := range user.RoleHistory() {
+			role, err := h.db.Authenticator(h.ctx()).GetRole(roleName)
+			if err != nil {
+				return err
+			}
+			if role == nil {
+				continue
+			}
+			if _, skipMapInit := resp.AdminRoleGrants[roleName]; !skipMapInit && roleHist.Source == channels.AdminGrant {
+				resp.AdminRoleGrants[roleName] = make(map[string]map[string]auth.GrantHistory)
+				resp.AdminRoleGrants[roleName][defaultKeyspace] = make(map[string]auth.GrantHistory)
+			} else if _, skipMapInit := resp.DynamicRoleGrants[roleName]; !skipMapInit && roleHist.Source == channels.DynamicGrant {
+				resp.DynamicRoleGrants[roleName] = make(map[string]map[string]auth.GrantHistory)
+				resp.DynamicRoleGrants[roleName][defaultKeyspace] = make(map[string]auth.GrantHistory)
+			}
+			for channel, chanEntry := range role.Channels() {
+				if channel == channels.DocumentStarChannel {
+					continue
+				}
+				roleChanHistory := roleHist
+				if chanEntry.VbSequence.Sequence > roleHist.Entries[len(roleHist.Entries)-1].StartSeq {
+					roleChanHistory.Entries = append(roleChanHistory.Entries, auth.GrantHistorySequencePair{StartSeq: chanEntry.VbSequence.Sequence, EndSeq: roleHist.Entries[len(roleHist.Entries)-1].EndSeq})
+				}
+				if roleHist.Source == channels.DynamicGrant {
+					if entry, ok := resp.DynamicRoleGrants[roleName][defaultKeyspace][channel]; ok {
+						roleChanHistory.Entries = append(entry.Entries, roleChanHistory.Entries...)
+					}
+				} else {
+					if entry, ok := resp.AdminRoleGrants[roleName][defaultKeyspace][channel]; ok {
+						roleChanHistory.Entries = append(entry.Entries, roleChanHistory.Entries...)
+					}
+				}
+				resp.addRoleGrants(roleName, roleHist.Source, defaultKeyspace, channel, roleChanHistory)
+			}
+			// loop over previous role channels
+			for channel, chanHistory := range role.ChannelHistory() {
+				base.InfofCtx(h.ctx(), base.KeyAll, "chanHistory", chanHistory.Source, channel)
```

Comment:
```suggestion
				base.InfofCtx(h.ctx(), base.KeyAll, "chanHistory", chanHistory.Source, channel)
```

drop debugging here?